### ISSUE TITLE
Failed HTTPS certificate negotiation shows as "error"

### ISSUE THREAD

**Issue by gwire** [openrightsgroup/cmp-issues#107]
A check of https://www.blocked.org.uk/results?url=https://www.ccc.de shows "error" for every check.

I'm assuming this is because they don't use an https certificate signed by a CA found in stock root (in the case of CCC they have their own CA).  If so, I assume expired certs will also produce the error.

It's probably an edge-case at the moment, thus low priority, but is there any way to
a) expose to the user that the error is due to a failed TLS negotiation
b) use some kind of observatory for checking the certificate is not ISP specific (ooni?)

**Comment by dantheta** [comment]
I've added a configurable which allows SSL certificate validation to be disabled.  When configured, a successful connection to a server with SSL will give an "OK" status, even if the cert is not verifiable.

A successful TCP connection that fails to start an SSL session will give a status of "SSLERROR".  The usual status responses apply to failed TCP connections ("ERROR" and "TIMEOUT").

**Comment by graphiclunarkid** [comment]
This sounds like a sensible solution. To go any further would require some checking outside the probe's individual environment. We discussed doing a pre-check of URLs via the server for a different reason some time ago however we decided not to at that time. Perhaps we should now discuss resurrecting the idea for this specific purpose? It would allow us to distinguish between genuine issues with a site and BT's blocking proxies not supporting TLS (for example).

**Comment by dantheta** [comment]
I've had a few thoughts about how to do a multi-network test in an elegant and scalable way.  At the moment, when a site is submitted, we post the URL to all of the ISP queues.  

What we could do instead (with only a little reconfiguration) is only post to a clean-line queue at that stage. When the result comes back from a clean line, the URL is posted to all of the other queues, along with the clean line result.  The probes running against the filtered line would be able to run their usual test, but with knowledge of what the clean line result was.  

Pro:
- We wouldn't be running any additional tests
- We can have multiple clean lines attached to the clean-line queue (reducing the possibility of bottleneck or SPOF)

Con:
- We might get false positives due to 'net issues that are only impacting one of the filtered-line probes.  
- If we lose all/most of the clean-lines, the throughput of the system is impacted, and may be halted.
- Result time goes up, since it now has to wait for 2 tests to run serially (the clean line and then the dirty line)

Any thoughts?

**Comment by graphiclunarkid** [comment]
Does (e.g.) http://downforeveryoneorjustme.com/ have an API we could use for the clean-line checks?

We could display the clean- and test-line checks separately so people might spot probe problems.

Do the requests have to be sequential or could we do them in parallel, but assume the clean test will pass unless proven otherwise, and if it fails push an AJAX update to the results table? That would mean the result would go from something like "error" -> "blocked (no response but site not down)" in cases where a slow clean-line test subsequently passes.

**Comment by dantheta** [comment]
I can't see any details for an API, but there are other similar systems which might.  Co-opting an external uptime monitoring services to run as a clean-line probe is an interesting idea.

Running the clean & filtered tests non-sequentially and gathering results could get quite complicated and ,getting the processing to be bulletproof in order to display reliably could be quite a challenge.  

If a slow clean-line test failed, it would have to update all of the already-received filtered-line results from "blocked" to "error", and also remember this for future filtered-line results that haven't arrived yet.  

I'd previously considered the comparison of different ISP results for a URL to be firmly in the realm of offline-processing, because it's so fiddly to do in real time (especially since the server doesn't know which probes are live or how many results to expect).  

Although the pipelined clean-then-filtered approach does increase the time-to-first-result, it's architecturally a lot simpler and should scale better than a process which has to wait on results at the server and gather/compare.  

We can certainly experiment with both, but I'd be keen to explore the simpler option first :)

**Comment by dantheta** [comment]
The probe and config update has been rolled out, and https://www.ccc.de is now shown as "OK" on all ISPs.  We no longer require SSL certs to be verifiable against the stock CA certs list.

**Comment by graphiclunarkid** [comment]
Seems like you've got a solution to the immediate issue so we can close this issue now. If you feel like exploring what checking URLs outside the probe environment could do for us I think we should raise a separate issue.

Thanks @dantheta!
